doc/development/style_guide.md

```diff
@@ -0,0 +1,211 @@
+  the `export ... show ...` statement. For example:
+  ```dart
+  export 'src/effects/provider_interfaces.dart'
+    show
+      AnchorProvider,
+      AngleProvider,
+      PositionProvider,
+      ScaleProvider,
+      SizeProvider;
+  ```
+
+
+### Assertions
+
+Use `assert`s to detect contract violations, or pre-condition/post-condition failures. Sometimes,
+however, using exceptions would be more appropriate. The following rules of thumb apply:
+
+- Use an assert with a clear error message to check for a condition that is in developers' control.
+  For example, when creating a component that takes an `opacity` level as an input, you should check
+  whether the value is in the range from 0 to 1. Consider also including the value itself into the
+  error message, to make it easier for the developer to debug the error:
+  ```dart
+  assert(0 <= opacity && opacity <= 1, 'The opacity value must be from 0 to 1: $opacity');
+  ```
+  Always use asserts as early as possible to detect possible violations. For example, check the
+  validity of `opacity` in the constructor/setter, instead of in the render function.
+
+  When adding such an assert, also include a test that checks that the assert triggers. This test
+  would verify that the component does not accept invalid input, and that the error message is what
+  you expect it to be.
+
+- Use an assert without an error message to check for a condition that cannot be triggered by the
+  developer through any means known to you. If such an assert does trigger, it would indicate a bug
+  in the Flame framework.
+
+  Such asserts serve as "mini-tests" directly in the code, and protect against future refactorings
+  that could create an erroneous internal state. It should not be possible to write a test that
+  would deliberately trigger such an assert.
+
+- Use an explicit if-check with an exception to test for a condition that may be outside of the
+  developer's control (i.e. it may depend on the environment or on user's input). When deciding
+  whether to use an assert or exception, consider the following question: is it possible for the
+  error condition to occur in production even after the developer has done extensive testing on
+  their side?
+
+
+### Class structure
+
+- Consider putting all class constructors at the top of the class. This makes it much easier to see
+  how the class ought to be used.
+
+- Try to make as much of your class' API private as possible, do not expose members "just in case".
+  This makes it much easier to modify/refactor the class later without it being a breaking change.
+
+  Remember to document all your public members! Documenting things is harder than it looks, and one
+  way to avoid the burden of documentation is to make as many variables private as possible.
+
+- If a class exposes a `List<X>` or `Vector2` property -- it is **NOT** an invitation to modify
+  them at will! Consider such properties as read-only, unless the documentation explicitly says that
```

Comment:
A lot of the time we actually want the users to modify the `Vector2`s directly? So that they don't need to create new objects just to update the value.